src/raw_node.rs

```diff
@@ -349,94 +384,290 @@ impl<T: Storage> RawNode<T> {
+                if !record.messages.is_empty() {
+                    self.messages.push(record.messages);
+                }
+            }
+        }
+
+        let ss = raft.soft_state();
+        if ss != self.prev_ss {
+            rd.ss = Some(ss);
+        }
+        let hs = raft.hard_state();
+        if hs != self.prev_hs {
+            if hs.vote != self.prev_hs.vote || hs.term != self.prev_hs.term {
+                rd.must_sync = true;
+            }
+            rd.hs = Some(hs);
+        }
+
+        if !raft.read_states.is_empty() {
+            rd.read_states = mem::take(&mut raft.read_states);
+        }
+
+        if raft.raft_log.unstable.snapshot.is_some() {
+            rd.snapshot = raft.raft_log.unstable.snapshot.clone().unwrap();
+            rd_record.snapshot = rd.snapshot.clone();
+            assert!(self.commit_since_index < rd.snapshot.get_metadata().index);
+            self.commit_since_index = rd.snapshot.get_metadata().index;
+            // If there is a snapshot, the latter entries can not be persisted
+            // so there is no committed entries.
+            assert!(
+                !raft
+                    .raft_log
+                    .has_next_entries_since(self.commit_since_index),
+                "has snapshot but also has committed entries since {}",
+                self.commit_since_index
+            );
+            self.pending_snapshot = true;
+            rd.must_sync = true;
+        } else {
+            rd.committed_entries = raft
+                .raft_log
+                .next_entries_since(self.commit_since_index)
+                .unwrap_or_default();
+            // Update raft uncommitted entries size
+            raft.reduce_uncommitted_size(&rd.committed_entries);
+            if let Some(e) = rd.committed_entries.last() {
+                assert!(self.commit_since_index < e.get_index());
+                self.commit_since_index = e.get_index();
+            }
+        }
+        // `unstable.stable_all` will take all unstable entries.
+        // So it should be called after `next_entries_between`.
+        rd.entries = raft.raft_log.unstable.stable_all();
+        if let Some(e) = rd.entries.last() {
+            // If the last entry exists, the entries must not empty, vice versa.
+            rd.must_sync = true;
+            rd_record.last_entry = Some((e.get_index(), e.get_term()));
+        }
+
+        if !self.messages.is_empty() {
+            rd.messages = mem::take(&mut self.messages);
+        }
+        if !raft.msgs.is_empty() {
+            if raft.state == StateRole::Leader {
+                // Leader can send messages immediately to make replication concurrently.
+                // For more details, check raft thesis 10.2.1.
+                rd.messages.push(mem::take(&mut raft.msgs));
+            } else {
+                rd_record.messages = mem::take(&mut raft.msgs);
+            }
+        }
+        self.records.push_back(rd_record);
```

Comment:
`commit_ready` has a number check for ensuring the order of function call, i.e. should call the `advance`-like function after calling `ready`. I think the cost is little because, in general, users should call `has_ready` before calling `ready`.

Reply:
> commit_ready has a number check for ensuring the order of function call, i.e. should call the advance-like function after calling ready

We can do the number check by replacing `Ready::number: u64` with `Ready::record_number: Option<u64>` and only set the `Ready::record_number` to `Some` when there are corresponding `ReadyRecord`, and do the checking:
```
if let Some(number) = rd.record_number {
        let rd_record = self.records.back().unwrap();
        assert!(rd_record.number == number);
}
```